### ISSUE TITLE
Volume setting in Android 9

### ISSUE THREAD

**Issue by mase76** [renyuneyun/Easer#217]
Hi!
(Bluetooth) volume setting does not work any more in Android 9.

**Comment by mase76** [comment]
Correction: BT volume setting works, but not higher than about 60%.

**Comment by renyuneyun** [comment]
I see a potential reason: Android 9 (or 8.1 already?) will notify the user when trying to increase the volume of a headset (including bluetooth headsets and wired headsets) larger than a certain limit (which is roughly 60-70 IIRC).

I'll try to figure out if there is a way around.

**Comment by mase76** [comment]
I know this warning from wired headsets only, not bluetooth. And the warning appears only once. Android 8 had this warning, too. But volume setting via easer to 100% was no problem.

**Comment by renyuneyun** [comment]
You are right. Only wired headsets got the warning.
I'll try to investigate.

By the way, does it work (i.e. change the volume you are supposed to listen to) if you change the "media" volume when connected to bluetooth headset?

**Comment by mase76** [comment]
Yes, this works. Even when using the headset volume buttons to change the bt media volume.

**Comment by renyuneyun** [comment]
Ok, I think now I know why...

Generally, we should avoid setting "bluetooth volume", but to set "volume".
On my device, setting media volume (with bluetooth connected) will change the corresponding bluetooth volume.

It seems Android has combined internal (e.g. speaker) and external (e.g. headset) volumes, making them (kind of) uniformed from the API. I was aware the bluetooth volume thing is actually a trick/workaround, but didn't imagine the maximum volume would be different.

**Comment by mase76** [comment]
I tried setting up to change the media volume. But on BT connection this does not affect the BT volume.

**Comment by renyuneyun** [comment]
Hmm, interesting...
Which ROM are you using? Maybe that matters?
(FYI, I'm using Oxygen OS 9.0.)

**Comment by mase76** [comment]
I am using Omnirom 9.0.0. But it seems to store the volume of every connected bt device.

**Comment by renyuneyun** [comment]
I just searched and got a little more information (not solving the problem though): Android 9 remembers individual (bluetooth) device's own volume, so your (phone's) behaviour should be the "correct" one. (And OxygenOS seems to have removed/hidden this somehow?)

Does this have anything to do with delays? For example, what if you adjust the volume in Easer after (e.g. 3 seconds after) connected to your bluetooth device?

**Comment by mase76** [comment]
The delay was to override the volume change done by the os. That was about half if I remember right. It seems, that the bt volume setting by Easer is not necessary any more. But I am not 100% sure.
